File: remote/src/main/java/com/kelsos/mbrc/interactors/PlayerStateInteractorImpl.java
package com.kelsos.mbrc.interactors;

import android.text.TextUtils;
import com.google.inject.Inject;
import com.kelsos.mbrc.annotations.PlayerState;
import com.kelsos.mbrc.cache.PlayerStateCache;
import com.kelsos.mbrc.dto.player.PlayState;
import com.kelsos.mbrc.services.api.PlayerService;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class PlayerStateInteractorImpl implements PlayerStateInteractor {
  @Inject private PlayerStateCache cache;
  @Inject private PlayerService service;

  @Override public Observable<String> getState() {
    Observable<String> networkRequest = service.getPlayState().map(PlayState::getValue).doOnNext(cache::setPlayState);
    Observable<String> cached = Observable.just(cache.getPlayState());

    return Observable.concat(networkRequest, cached)
        .filter(s -> !TextUtils.isEmpty(s) && !PlayerState.UNDEFINED.equals(s))
        .doOnError(throwable -> Observable.just(PlayerState.STOPPED))
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread());
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/interactors/PlayerStateInteractorImpl.kt
package com.kelsos.mbrc.interactors

import com.kelsos.mbrc.annotations.PlayerState
import com.kelsos.mbrc.cache.PlayerStateCache
import com.kelsos.mbrc.extensions.task
import com.kelsos.mbrc.services.api.PlayerService
import rx.Observable
import rx.lang.kotlin.toSingletonObservable
import javax.inject.Inject

class PlayerStateInteractorImpl
@Inject constructor(private val cache: PlayerStateCache,
                    private val service: PlayerService) : PlayerStateInteractor {

  override fun getState(): Observable<String> {
    val networkRequest = service.getPlayState()
        .map { it.value }
        .doOnNext { cache.playState = it }

    val cached = cache.playState.toSingletonObservable()

    return Observable.concat(networkRequest, cached)
        .filter { PlayerState.UNDEFINED != it }
        .doOnError { PlayerState.STOPPED.toSingletonObservable() }
        .firstOrDefault(PlayerState.STOPPED)
        .task()
  }
}
